internal/providers/kubevirt/kubevirt.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2021 Red Hat.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The KubeVirt (https://kubevirt.io) provider fetches
+// configuration from the userdata available in the config-drive. It is similar
+// to Openstack and uses the same APIs.
+
+package kubevirt
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/coreos/ignition/v2/config/v3_4_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	ut "github.com/coreos/ignition/v2/internal/util"
+
+	"github.com/coreos/vcontext/report"
+)
+
+const (
+	configDriveUserdataPath = "/openstack/latest/user_data"
+)
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	data, err := fetchConfigFromDevice(f.Logger, filepath.Join(distro.DiskByLabelDir(), "config-2"))
+	if err != nil {
+		return types.Config{}, report.Report{}, err
+	}
+
+	return util.ParseConfig(f.Logger, data)
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return (err == nil)
+}
+
+func fetchConfigFromDevice(logger *log.Logger, path string) ([]byte, error) {
+	// The config drive is always attached, even if there's no user-data passed
```

Comment:
Can you confirm that this code correctly detects a missing Ignition config if the machine is launched without one?

Reply:
Updated the code and the test results above. In fact kubevirt does not always provide a config-drive. Ignition will now retry a few times and then continue with an "empty" config. I hope this is the desired behaviour from the ignition perspective.

Reply:
Uh oh, that's going to be a problem.  What are the circumstances in which kubevirt doesn't provide a config drive?  Is there any way to force it to do so?  Ignition can't safely time out waiting for a config, since that creates a race condition: a node could boot successfully without having applied the requested configuration.

Reply:
That would work.  So the user can select whether to use a config drive?  Is it possible for the user to intentionally provide a config drive but _not_ an Ignition config?  If so, we should confirm that this PR correctly handles that case.

Reply:
Excellent test logs, thank you!